core/primitives/src/congestion_info.rs

```diff
@@ -150,13 +153,23 @@ impl CongestionInfo {
     // information from the chunk extra.
     //
     // TODO(congestion_control) validate allowed shard
-    pub fn validate_extra_and_header(extra: &CongestionInfo, header: &CongestionInfo) -> bool {
+    pub fn validate_extra_and_header(
+        extra_protocol_version: ProtocolVersion,
+        extra: &CongestionInfo,
+        header: &CongestionInfo,
+    ) -> bool {
         match (extra, header) {
             (CongestionInfo::V1(extra), CongestionInfo::V1(header)) => {
+                let correct_allowed_shard =
+                    if ProtocolFeature::CongestionControl.enabled(extra_protocol_version) {
+                        extra.allowed_shard == header.allowed_shard
+                    } else {
+                        true
```

Comment:
Shouldn't this at least check that the shard id is a valid shard id for the current shard layout?

Reply:
In theory yes, in practice no, since this code path will never be executed in production and I doubt anyone would want to attack this in statelessnet. It won't be executed in production because it should be deleted all together when congestion control is stabilized. Again this mess is here only because of statelessnet.

Reply:
I see, makes sense!